src/utils/pgwire/src/pg_extended.rs

```diff
@@ -20,19 +20,47 @@ use regex::Regex;
 use crate::pg_field_descriptor::{PgFieldDescriptor, TypeOid};
 use crate::pg_protocol::cstr_to_str;
 
-fn replace_params(query_string: String, generic_params: &[usize], params: &[Bytes]) -> String {
+fn replace_params(query_string: String, generic_params: &[usize], params: &[String]) -> String {
     let mut tmp = query_string;
+    tmp.push(' ');
     for &i in generic_params.iter() {
         let pattern = Regex::new(format!(r"(?P<x>\${})(?P<y>[,;\s]+)", i).as_str()).unwrap();
-        let param = cstr_to_str(&params[i.sub(1)]).unwrap();
+        let param = &params[i.sub(1)];
         tmp = pattern
             .replace_all(&tmp, format!("{}$y", param))
             .to_string();
     }
-
+    tmp.pop();
```

Comment:
Why here extra pop?

Reply:
In the line 25, we have tmp.push(' '). 

This ' ' helps us to handle the specific case, such as: 
**'SELECT $2,$1'** 
Due to '$1' being the last string without any punctuation behind it, it can't be matched by the regex.
To simplify this edge example, I push ' ' to make it be 'SELECT $2,$1 ', now '$1 ' can be matched.

When I finish processing, I pop the final ' '.

Reply:
I find another way that is easier to understand.